pkg/server/commands/listusers/list_users_rpc_test.go

```diff
@@ -2,301 +2,1245 @@ package listusers
 				tuple.NewTupleKey("group:eng", "member", "group:fga#member"),
-				tuple.NewTupleKey("group:fga", "member", "user:jon"),
+				tuple.NewTupleKey("document:1", "viewer", "group:eng#member"),
+				tuple.NewTupleKey("document:1", "viewer", "group:other#member"),
 			},
-			expectedUsers: []*openfgav1.User{
-				{
-					User: &openfgav1.User_Object{
-						Object: &openfgav1.Object{Type: "user", Id: "jon"},
+			expectedUsers: []string{"group:fga#member", "group:eng#member", "group:other#member"},
+		},
+		{
+			name: "userset_user_granularity_with_contextual_tuples",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "viewer",
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "user",
 					},
 				},
-				{
-					User: &openfgav1.User_Object{
-						Object: &openfgav1.Object{Type: "user", Id: "hawker"},
+				ContextualTuples: &openfgav1.ContextualTupleKeys{
+					TupleKeys: []*openfgav1.TupleKey{
+						tuple.NewTupleKey("group:marketing", "member", "user:jon"),
+						tuple.NewTupleKey("document:1", "viewer", "group:eng#member"),
 					},
 				},
 			},
+			model: model,
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("group:eng", "member", "user:will"),
+				tuple.NewTupleKey("group:eng", "member", "user:maria"),
+			},
+			expectedUsers: []string{"user:will", "user:maria"},
 		},
 		{
-			name: "rewritten_direct_relationship_through_computed_userset",
+			name:                  "tuple_defines_itself",
+			TemporarilySkipReason: "because it wants to return `document:1`",
 			req: &openfgav1.ListUsersRequest{
-				StoreId:              storeID,
-				AuthorizationModelId: modelID,
-				Object:               &openfgav1.Object{Type: "document", Id: "1"},
-				Relation:             "viewer",
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "viewer",
 				UserFilters: []*openfgav1.ListUsersFilter{
 					{
-						Type: "user",
+						Type: "document",
 					},
 				},
 			},
-			typeDefinitions: parser.MustParse(`
+			model: `model
+            schema 1.1
 			type user
 			type document
 			  relations
-			    define editor: [user] as self
-			    define viewer as editor
-			`),
+			    define viewer: [user]
+			`,
+			tuples:        []*openfgav1.TupleKey{},
+			expectedUsers: []string{},
+		},
+		{
+			name:                  "userset_defines_itself",
+			TemporarilySkipReason: "because reflexive relationships not supported yet",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "viewer",
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type:     "document",
+						Relation: "viewer",
+					},
+				},
+			},
+			model: `model
+            schema 1.1
+			type user
+			type document
+			  relations
+			    define viewer: [user]
+			`,
+			tuples:        []*openfgav1.TupleKey{},
+			expectedUsers: []string{"document:1#viewer"},
+		},
+	}
```

Comment:
will be an interesting case if you have the case where a user belongs to multiple group and multiple groups as viewer access to the document

For example]
```
model
	schema 1.1
type user
type group
	relations
		define member: [user]
type document
	relations
		define viewer: [ user, group#member ]			
```

with tuples
```
tuple.NewTupleKey("group:eng", "member", "user:will"),
tuple.NewTupleKey("group:eng", "member", "user:maria"),
tuple.NewTupleKey("group:core", "member", "user:will"),
tuple.NewTupleKey("document:1", "viewer", "group:core#member"),
tuple.NewTupleKey("document:1", "viewer", "group:eng#member"),
```
You should see `user:will` and `user:maria` appear once in the return.  (try to prevent `user:will` from appearing twice)

Reply:
Lack of deduplication can be observed in a few of the other test cases but this has standalone value. Just added.